src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -492,6 +557,81 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
             }
         }
 
+        private async Task StartWritingAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // We loop to this method from the callback.
+            // If the last chunk was just completed from async callback (count < 0), we complete user request.
+            if (count >= 0 )
+            {
+                byte[] outBuffer = null;
+                if (_pinnableOutputBufferInUse == null)
+                {
+                    if (_pinnableOutputBuffer == null)
+                    {
+                        _pinnableOutputBuffer = s_PinnableWriteBufferCache.AllocateBuffer();
+                    }
+
+                    _pinnableOutputBufferInUse = buffer;
+                    outBuffer = _pinnableOutputBuffer;
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
```

Comment:
@brianrob @vancem own this portion in NetFX. Since it needs to be enabled I don't see the noise and the cost of the single if should not be significant to justify a change.